solidity/random-beacon/contracts/libraries/Groups.sol

```diff
@@ -6,27 +6,43 @@ pragma solidity ^0.8.6;
 ///      groups creation (only one group creation happens at a time). A candidate
 ///      group has to be popped or activated before adding a new candidate group.
 library Groups {
-    event CandidateGroupRegistered(bytes indexed groupPubKey);
-
-    event CandidateGroupRemoved(bytes indexed groupPubKey);
-
-    event GroupActivated(uint64 indexed groupId, bytes indexed groupPubKey);
-
     struct Group {
         bytes groupPubKey;
-        uint256 activationTimestamp;
+        uint256 activationBlockNumber;
         uint32[] members;
+        // When selected group does not create a relay entry on-time it should
+        // be marked as terminated.
+        bool terminated;
     }
 
     struct Data {
         // Mapping of keccak256 hashes of group public keys to groups details.
         mapping(bytes32 => Group) groupsData;
         // Holds keccak256 hashes of group public keys in the order of registration.
         bytes32[] groupsRegistry;
+        // Group ids that were active but failed creating a relay entry. When an
+        // active-terminated group qualifies to become 'expired', then it will
+        // be removed from this array.
+        uint64[] activeTerminatedGroups;
         // TODO: Remember about decreasing the counter in case of expiration or termination.
         uint64 activeGroupsCount;
+        // Points to the first active group, it is also the expired groups counter.
+        uint64 expiredGroupOffset;
+        // Group lifetime in blocks. When a group reached its lifetime, it
+        // is no longer selected for new relay requests but may still be
+        // responsible for submitting relay entry if relay request assigned
+        // to that group is still pending.
+        uint256 groupLifetime;
+        // Calculated in the Relay.sol lib.
+        uint256 relayEntryTimeout;
```

Comment:
Bear in mind this value may change. This is because both `relayEntrySubmissionEligibilityDelay` and `relayEntryHardTimeout` are governable. I see we're now setting `relayEntryTimeout` only in `RandomBeacon` constructor which is not enough.